C++/Graphs/Dinic.cpp

```diff
@@ -0,0 +1,227 @@
+            {
+                // Level of current vertex is,
+                // level of parent + 1
+                level[e.v] = level[u] + 1;
+
+                q.push_back(e.v);
+            }
+        }
+    }
+
+    // IF we can not reach to the sink we
+    // return false else true
+    return level[t] < 0 ? false : true;
+}
+
+// A DFS based function to send flow after BFS has
+// figured out that there is a possible flow and
+// constructed levels. This function called multiple
+// times for a single call of BFS.
+// flow : Current flow send by parent function call
+// start[] : To keep track of next edge to be explored.
+//		 start[i] stores count of edges explored
+//		 from i.
+// u : Current vertex
+// t : Sink
+int Graph::sendFlow(int u, int flow, int t, int start[])
+{
+    // Sink reached
+    if (u == t)
+        return flow;
+
+    // Traverse all adjacent edges one -by - one.
+    for (; start[u] < adj[u].size(); start[u]++)
+    {
+        // Pick next edge from adjacency list of u
+        Edge &e = adj[u][start[u]];
+
+        if (level[e.v] == level[u] + 1 && e.flow < e.C)
+        {
+            // find minimum flow from u to t
+            int curr_flow = min(flow, e.C - e.flow);
+
+            int temp_flow = sendFlow(e.v, curr_flow, t, start);
+
+            // flow is greater than zero
+            if (temp_flow > 0)
+            {
+                // add flow to current edge
+                e.flow += temp_flow;
+
+                // subtract flow from reverse edge
+                // of current edge
+                adj[e.v][e.rev].flow -= temp_flow;
+                return temp_flow;
+            }
+        }
+    }
+
+    return 0;
+}
+
+// Returns maximum flow in graph
+int Graph::DinicMaxflow(int s, int t)
+{
+    // Corner case
+    if (s == t)
+        return -1;
+
+    int total = 0; // Initialize result
+
+    // Augment the flow while there is path
+    // from source to sink
+    while (BFS(s, t) == true)
+    {
+        // store how many edges are visited
+        // from V { 0 to V }
+        int *start = new int[V + 1]{0};
+
+        // while flow is not zero in graph from S to D
+        while (int flow = sendFlow(s, INT_MAX, t, start))
+        {
+            // Add path flow to overall flow
+            total += flow;
+        }
+
+        // Remove allocated array
+        delete[] start;
+    }
+
+    // return maximum flow
+    return total;
+}
+
+// Driver Code
+int main()
+{
+    cout << "Enter the number of nodes in the graph : " << endl;
+    int n;
+    cin >> n;
+    Graph g(n);
+    int m;
+    cout << "Enter the number of edges in the graph : " << endl;
+    cin >> m;
+
+    cout << "Enter the input for each edge in the form u , v, C where u = starting node , v = end node and C = capacity : " << endl;
+    int u, v, c;
+    for (int i = 0; i < n; i++)
+    {
+        cin >> u, v, c;
+        g.addEdge(u, v, c);
+    }
+    // g.addEdge(0, 1, 16);
```

Comment:
Why have you commented these lines

Reply:
> Why have you commented these lines

I have given sample graph examples to the user for reference.